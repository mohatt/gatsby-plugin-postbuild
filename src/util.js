import path from 'path'
import crypto from 'crypto'
import _ from 'lodash'
import Debug from 'debug'

/**
 * Global options object
 *
 * @type {Object}
 */
export let options = {
  _plugin: 'gatsby-plugin-postbuild'
}

/**
 * Initializes plguin options and other utilities
 *
 * @param {Object} $0
 * @param   {Object} $0.defaultOptions - Options defaults
 * @param   {Object} $0.pluginOptions - Plugin options
 * @param   {Object} $0.gatsby - Gatsby's onPreBootstrap event object
 */
export function bootstrap ({ defaultOptions, pluginOptions, gatsby }) {
  // Merge user-defined options with defaults
  options = _.mergeWith(
    options,
    defaultOptions,
    pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => _.isArray(from) ? from : undefined
  )
  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap = {
    10000: {
      text: context => `${options._plugin} threw an error while running:\n ${context.message}`,
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (reporter.setErrorMap) {
    reporter.setErrorMap(errorMap)
  }
  debug('Plugin initialized', options)
}

/**
 * Creates a new structured error to pass to gatsby's reporter
 *
 * @constructor
 * @param {string|Object} message
 * @param {Error|null} error
 * @param {number} code
 * @return {{ id: string, context: {Object}, error: error}}
 */
export function ReporterError (message, error = null, code = 10000) {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error && typeof error === 'object'
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 *
 * @param {string} message
 * @param {any} data
 */
export function debug (message, data = null) {
  createDebug()(...arguments)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace = '') {
  let ns = options._plugin
  if (namespace) {
    ns += ':' + namespace
  }

  createDebug.cache ??= new Map()
  if (createDebug.cache.has(ns)) {
    return createDebug.cache.get(ns)
  }

  const realDebug = Debug(ns)
  const debug = function (message) {
    arguments[0] = arguments.length > 1
      ? `${message} %O \n\n`
      : `${message} \n\n`
    realDebug(...arguments)
  }

  createDebug.cache.set(ns, debug)
  return debug
}

/**
 * Creates a sha1 hash from a string.
 *
 * @param {string} data
 * @param {boolean} base64
 */
export function sha1 (data, base64 = false) {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}
