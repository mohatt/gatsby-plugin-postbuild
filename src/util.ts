import path from 'path'
import crypto from 'crypto'
import { mergeWith, isArray } from 'lodash'
import Debug from 'debug'
import { defaults, IPluginOptions } from './options'
import {
  GatsbyNodeHelpers,
  GatsbyPluginOptions,
  GatsbyReporterErrorMap,
  GatsbyReporterErrorMeta
} from './gatsby'

/**
 * Options interface
 */
export interface IOptions extends IPluginOptions {
  _plugin: string
  _root: string
  _public: string
  _pathPrefix: string
}

/**
 * Global options object
 */
export let options: IOptions = {
  _plugin: 'gatsby-plugin-postbuild',
  _root: '',
  _public: '',
  _pathPrefix: '',
  ...defaults
}

/**
 * Initializes plguin options and other utilities
 */
export function bootstrap ({ pluginOptions, gatsby }: {
  pluginOptions: GatsbyPluginOptions
  gatsby: GatsbyNodeHelpers
}): void {
  // Merge user-defined options with defaults
  options = mergeWith(options, pluginOptions,
    // Ensure arrays aren't merged by index
    (to, from) => isArray(from) ? from : undefined
  )

  // set private options
  options._root = gatsby.store.getState().program.directory
  options._public = path.join(options._root, 'public')
  options._pathPrefix = gatsby.pathPrefix

  // Initialize reporter
  const reporter = gatsby.reporter
  const errorMap: GatsbyReporterErrorMap = {
    10000: {
      text: (context: { message: string }) => {
        return `${options._plugin} threw an error while running:\n ${context.message}`
      },
      level: 'ERROR',
      type: 'PLUGIN'
    }
  }
  if (typeof reporter.setErrorMap === 'function') {
    reporter.setErrorMap(errorMap)
  }

  // Debug final options
  debug('Plugin initialized', options)
}

/**
 * Creates a structured error object to pass to gatsby's reporter
 */
export function createGatsbyError (message: string|Object, error?: Error, code = 10000): GatsbyReporterErrorMeta {
  return {
    id: `${options._plugin}_${code}`,
    context: typeof message === 'object'
      ? message
      : { message },
    error: error instanceof Error
      ? error
      : undefined
  }
}

/**
 * Prints a debug message
 */
export function debug (message: string, ...params: any[]): void {
  createDebug()(message, ...params)
}

/**
 * Creates a debug instance for the given namespace
 *
 * @param {string} namespace
 * @return {Function}
 */
export function createDebug (namespace: string = ''): Function {
  let ns = options._plugin
  if (namespace !== '') {
    ns += ':' + namespace
  }

  const debug = Debug(ns)
  return function (message: string, ...params: any[]) {
    message = params.length > 0
      ? `${message} %O \n\n`
      : `${message} \n\n`
    debug(message, ...params)
  }
}

/**
 * Creates a sha1 hash from a string.
 */
export function sha1 (data: string, base64 = false): string {
  return crypto.createHash('sha1')
    .update(data, 'utf8')
    .digest(base64 ? 'base64' : 'hex')
}

/**
 * Returns the file extension or checks if
 * a file has a given extension
 */
export function extName (file: string, checkExt?: string): string|boolean {
  const ext = path.extname(file).replace('.', '').toLowerCase()
  if (checkExt === undefined) return ext
  return ext === checkExt
}
