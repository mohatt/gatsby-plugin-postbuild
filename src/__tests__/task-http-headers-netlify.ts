import { testE2E, testHeaders } from '@test/util'

describe('task: http-headers [netlify]', () => {
  it('works correctly', async () => {
    await testE2E('task: http-headers [netlify]', {
      'http-headers': {
        enabled: true,
        provider: 'netlify',
        headers: testHeaders,
      },
      ignore: ['extra.html'],
    })
  })
})
