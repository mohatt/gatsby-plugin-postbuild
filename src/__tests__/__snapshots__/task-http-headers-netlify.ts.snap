// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`task: http-headers [netlify] > works correctly > activity.panic 1`] = `[]`;

exports[`task: http-headers [netlify] > works correctly > activity.setStatus 1`] = `
[
  [
    "Loaded 1/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 2/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 3/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 4/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 5/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 6/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 7/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 8/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 9/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 0/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 1/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 2/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 3/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 4/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 5/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 6/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 7/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 0/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 1/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 2/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 3/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 4/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 5/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 6/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 7/11",
  ],
  [
    "Loaded 10/11 Processed 8/11 Wrote 8/11",
  ],
  [
    "Loaded 10/11 Processed 9/11 Wrote 8/11",
  ],
  [
    "Loaded 10/11 Processed 10/11 Wrote 8/11",
  ],
  [
    "Loaded 11/11 Processed 10/11 Wrote 8/11",
  ],
  [
    "Loaded 11/11 Processed 10/11 Wrote 9/11",
  ],
  [
    "Loaded 11/11 Processed 10/11 Wrote 10/11",
  ],
  [
    "Loaded 11/11 Processed 11/11 Wrote 10/11",
  ],
  [
    "Loaded 11/11 Processed 11/11 Wrote 11/11",
  ],
  [
    "Done processing 11 files saving a total of 3.01kB",
  ],
]
`;

exports[`task: http-headers [netlify] > works correctly > info 1`] = `
[
  [
    "[gatsby-plugin-postbuild]: Report
 update 404.html [6.11kB -1%]
 update blog/index.html [10.13kB -2%]
 update 404/index.html [6.11kB -1%]
 update blog/blog-post-3/index.html [7.3kB -1%]
 update blog/blog-post-2/index.html [7.3kB -1%]
 update blog/blog-post-1/index.html [7.3kB -1%]
 update blog/tag/tag3/index.html [9.43kB -2%]
 update blog/tag/tag2/index.html [9.43kB -2%]
 update index.html [6.54kB -4%]
 update about/index.html [12.82kB -7%]
 update blog/tag/tag1/index.html [9.43kB -2%]
 create _headers [1kB]
 create postbuild.log.json [1.48kB]",
  ],
]
`;

exports[`task: http-headers [netlify] > works correctly > panic 1`] = `[]`;

exports[`task: http-headers [netlify] > works correctly > verbose 1`] = `[]`;
